Introduce some policy constants from Bitcoin Core

darosior opened the pull request:
~~Not sure about the place since all other constants are consensus, but it seemed awkward to have it eg in the `transaction` module while there is a `constants` one :shrug:~~

EDIT: this now introduces a minimal set of policy stuff from `bitcoind` in a new `policy` module.

stevenroose commented:
Hmm I'm not sure how I feel about standardness constants in here because they depend on Bitcoin Core versions, so they might get outdated. Perhaps we can have a `mod standard` with such a disclaimer placed on it or so?

TheBlueMatt commented:
> Hmm I'm not sure how I feel about standardness constants in here because they depend on Bitcoin Core versions, so they might get outdated

There are effectively multiple different types of standardness rules - there are those that depend on Bitcoin Core versions and are likely to be relaxed as a part of a future soft-fork, and those which are effectively anti-DoS rules and are unlikely to be made more strict (though may be relaxed with algorithmic improvements). For better or for worse, things like the max standard weight and dust limits are now a part of the lightning trust model, which secures a nontrivial amount of funds. Changing them to be more strict in the future would imply a potential funds loss vulnerability in lightning.

dr-orlovsky commented on src/blockdata/constants.rs:
```diff
@@ -47,6 +47,8 @@ pub const MIN_TRANSACTION_WEIGHT: u32 = 4 * 60;
 pub const WITNESS_SCALE_FACTOR: usize = 4;
 
+/// Maximum weight of a transaction for it to be relayed by most nodes on the network
+pub const MAX_STANDARD_TRANSACTION_WEIGHT: u32 = 400_000;
```
Probably we should match Bitcoin Core name `MAX_STANDARD_TX_WEIGHT` since it was the policy defined by @apoelstra which we should continue to follow

https://github.com/bitcoin/bitcoin/blob/4a540683ec40393d6369da1a9e02e45614db936d/src/policy/policy.h#L24

dr-orlovsky reviewed (changes requested):
Probably outside of the scope of this PR we should try to add some consistentcy into which constants from Bitcoin Core we are moving to rust bitcoin and where they should reside.

For now, I have nothing against adding a single constant (it is better than none), but at least from here https://github.com/bitcoin/bitcoin/blob/4a540683ec40393d6369da1a9e02e45614db936d/src/policy/policy.h#L19-L54 I see a set of constants which can be added alongside `MAX_STANDARD_TX_WEIGHT`:
- `DEFAULT_BLOCK_MAX_WEIGHT`
- `DEFAULT_BLOCK_MIN_TX_FEE`
- `MIN_STANDARD_TX_NONWITNESS_SIZE`

darosior commented:
I agree, i'll spot the constants i find appropriate and will include them here. Would you conceptually accept this @apoelstra ?

sgeisler commented:
For the record, I like the idea of having a separate module for standardness constants with a disclaimer as proposed by @stevenroose. Iiuc that would also be a good place for the constants introduced in #566 @TheBlueMatt?

apoelstra commented:
concept ACK having this value in here. I agree that "in theory" it could change but in practice nobody has ever proposed to change it, and in future Taproot will make massive transactions even less useful.

darosior commented:
Added some constants that i found sensible, leaving the Script-related ones in `rust-miniscipt`, though we might want to include them here as well if non-Miniscript users want to access them too? `rust-miniscript` could access them from here once/if we have a `policy` module. Happy to introduce more if i missed some!

Renamed `MAX_STANDARD_TRANSACTION_WEIGHT` to `MAX_STANDARD_TX_WEIGHT` and added a counterpart to `bitcoind`'s `GetVirtualTransactionSize`, which is now possible since this introduces the sigop cost constant.

Overall i tried to use the same units as `bitcoind`, so it results in a lot of coercion (but only on constants!).

> For now, I have nothing against adding a single constant (it is better than none), but at least from here https://github.com/bitcoin/bitcoin/blob/4a540683ec40393d6369da1a9e02e45614db936d/src/policy/policy.h#L19-L54 I see a set of constants which can be added alongside `MAX_STANDARD_TX_WEIGHT`:
> 
>   * `DEFAULT_BLOCK_MAX_WEIGHT`
> 
>   * `DEFAULT_BLOCK_MIN_TX_FEE`
> 
>   * `MIN_STANDARD_TX_NONWITNESS_SIZE`

I don't think that it's sensible to add `DEFAULT_BLOCK_MAX_WEIGHT` and `DEFAULT_BLOCK_MIN_TX_FEE`. It's internal to Bitcoin Core, only used in the mining code (and we don't have any here?). Added `MIN_STANDARD_TX_NONWITNESS_SIZE`.

dr-orlovsky commented on src/policy/mod.rs:
```diff
@@ -0,0 +1,51 @@
+/// virtual size computation.
+pub const DEFAULT_BYTES_PER_SIGOP: u32 = 20;
+
+/// The minimum feerate, in sats per kilo-virtualbyte, for defining dust. An output is considered
```
```suggestion
/// The minimum feerate, in sats per virtual kilobyte, for defining dust. An output is considered
```

darosior replied:
Addressed.

dr-orlovsky commented on src/lib.rs:
```diff
@@ -68,6 +68,7 @@ pub mod blockdata;
 pub mod util;
 pub mod consensus;
 pub mod hash_types;
+pub mod policy;
```
Should we have this as a separate module, or it can just be a file inside `network`?

sgeisler replied:
I think this is fine, it's not really related to the p2p network layer imo, more general mempool policy.

dr-orlovsky reviewed (dismissed):
Checked against constant values in Bitcoin Core. General utACK with few nits.

sgeisler commented on src/blockdata/constants.rs:
```diff
@@ -46,7 +46,8 @@ pub const MAX_BLOCK_WEIGHT: u32 = 4_000_000;
 pub const WITNESS_SCALE_FACTOR: usize = 4;
-
+/// The maximum allowed number of signature check operations in a block
+pub const MAX_BLOCK_SIGOPS_COST: i64 = 80000;
```
Very tiny nit: why not put a thousands separator in there like `80_000`, makes counting zeroes easier.

darosior replied:
Very nit much wow! You are right, added it to all the constants introduced :).

sgeisler commented on src/policy/mod.rs:
```diff
@@ -0,0 +1,44 @@
+//!
+//! # *Warning*
+//! While the constants present in this module are very unlikely to change, they do not define
+//! Bitcoin. As such they must not be relied upon as if they were consensus rules.
```
Maybe a link to the core code would be nice here and the bitcoind version from which the constants were taken?

darosior replied:
Added the version and the tip

sgeisler commented on src/policy/mod.rs:
```diff
@@ -42,3 +43,9 @@ pub const DEFAULT_BYTES_PER_SIGOP: u32 = 20;
+
+/// The virtual transaction size, as computed by default by bitcoind node.
+pub fn get_virtual_tx_size(weight: i64, n_sigops: i64) -> i64 {
+    (cmp::max(weight, n_sigops * DEFAULT_BYTES_PER_SIGOP as i64) + WITNESS_SCALE_FACTOR as i64 - 1)
```
Idk enough about core, but it appears to have a more general version of this function where `DEFAULT_BYTES_PER_SIGOP` is variable. Is this actually never used?

darosior replied:
It's used in a rather convoluted manner in Bitcoin Core for the sigop cost to be configurable. We only care about the default here, which is what 99.9% of the network nodes run.

sgeisler reviewed (approved):
utACK 05a813ad6395702e30ae5f7d2ef6a0a5a235e2d2

Left some minor nits and questions. But nothing blocking if you want to go ahead and merge.

darosior commented:
Added two more constants that i forgot and are useful to contract and wallet developers: the default minimum relay feerate and the default mempool expiration.

sgeisler reviewed (approved):
utACK 7345aa60d9486dd4dffd8aa9e4d7fbbd52057a55

dr-orlovsky commented on src/policy/mod.rs:
```diff
@@ -0,0 +1,61 @@
+//! Policy
+//!
+//! This module exposes some constants and functions used in the reference implementation and which
+//! as a consequence defines some network rules.
```
NIT typo
```suggestion
//! as a consequence define some network rules.
```

dr-orlovsky commented on src/policy/mod.rs:
```diff
@@ -0,0 +1,61 @@
+/// mempools.
+pub const DEFAULT_MEMPOOL_EXPIRY: u32 = 336;
+
+/// The virtual transaction size, as computed by default by bitcoind node.
```
Nit (this is more clear description):
```suggestion
/// The virtual transaction size, as computed by Bitcoin Core `GetVirtualTransactionSize` using default bytes per sigop value
```

dr-orlovsky reviewed (approved):
ACK, re-checked against Bitcoin Core values. Two nits for comments which can be ignored. If not, pls add them as a separate commit so there will be no need to re-check constant values against Bitcoin Core

stevenroose reviewed (approved):
utACK

sgeisler commented:
3 ACKs, CI+1 local test run, minor API change: merged
